src/logic/hooks/useEstimateGasLimit.tsx

```diff
@@ -0,0 +1,43 @@
+import { useEffect, useState } from 'react'
+
+const DEFAULT_GAS_LIMIT = '0'
+
+export const useEstimateGasLimit = (
+  estimationFn: () => Promise<number>,
+  isExecution: boolean,
+  txData: string,
+  manualGasLimit: string | undefined,
+): string | undefined => {
+  const [gasLimit, setGasLimit] = useState<string | undefined>(manualGasLimit)
+
+  useEffect(() => {
+    let isCurrent = true
+
+    if (manualGasLimit) {
+      setGasLimit(manualGasLimit)
+      return
+    }
+
+    if (!isExecution || !txData) {
+      return
+    }
+
+    const estimateGasLimit = async () => {
+      try {
+        const estimation = await estimationFn()
+        isCurrent && setGasLimit(estimation.toString())
+      } catch (e) {
+        console.log(e)
+        setGasLimit(DEFAULT_GAS_LIMIT)
+      }
+    }
+
+    estimateGasLimit()
+
+    return () => {
+      isCurrent = false
+    }
+  }, [estimationFn, isExecution, manualGasLimit, txData])
```

Comment:
Can we use useAsync here?

Reply:
I tried that initially for both hooks but since they have to early return in case of some flags it didn't work with `useAsync`. It will always try to run the callback which we need to prevent in some cases i.e. if `gasLimit` is undefined or it is not an execution.

Reply:
It's fine to have an early return in the callback.
```
const [data, error, loading] = useAsync(async () => {
    if (!isExecution || !txData) return
     return await estimationFn()
}, [estimationFn, isExecution, manualGasLimit, txData])

useEffect(() => {
  if (manualGasLimit) {
    setGasLimit(manualGasLimit)
    return
  }
  if (data) setGasLimit(data.toString())
  if (error) setGasLimit(DEFAULT_GAS_LIMIT)
}, [manualGasLimit, data, error])

```

Reply:
Works very nicely with this approach and most tests held up the change.